.github/workflows/ci.yml

```
@@ -0,0 +1,146 @@
+on: [push, pull_request, release]
+name: Continuous Integration
+
+jobs:
+  test:
+    name: Test
+    runs-on: windows-latest
+    steps:
+        - name: Checkout
+          uses: actions/checkout@v2
+
+        - name: Install .NET 5.0.x
+          uses: actions/setup-dotnet@v1
+          with:
+             dotnet-version: '5.0.x'
+
+        - name: Compile
+          run: dotnet build -c Debug build/Desktop.proj
+
+        - name: Test
+          run: dotnet test $pwd/*.Tests/bin/Debug/*/*.Tests.dll --settings $pwd/build/vstestconfig.runsettings --logger:nunit
+          shell: powershell
+
+        - name: Report NUnit Logs
+          uses: MirrorNG/nunit-reporter@v1.0.9
+          with:
+            path: './TestResults/TestResults.xml'
+            access-token: ${{ secrets.GITHUB_TOKEN }}
+
+  inspect-code:
+     name: Code Quality
+     needs: test
+     runs-on: ubuntu-latest
+     steps:
+        - name: Checkout
+          uses: actions/checkout@v2
+
+        # FIXME: Tools won't run in .NET 5.0 unless you install 3.1.x LTS and 2.1 LTS side by side.
+        # https://itnext.io/how-to-support-multiple-net-sdks-in-github-actions-workflows-b988daa884e
+        - name: Install .NET 2.1.x LTS
+          uses: actions/setup-dotnet@v1
+          with:
+            dotnet-version: '2.1.x'
+
+        - name: Install .NET 3.1.x LTS
+          uses: actions/setup-dotnet@v1
+          with:
+             dotnet-version: '3.1.x'
+
+        - name: Install .NET 5.0.x
+          uses: actions/setup-dotnet@v1
+          with:
+             dotnet-version: '5.0.x'
+
+        - name: Restore Tools
+          run: dotnet tool restore
+
+        - name: Restore Packages
+          run: dotnet restore
+
+        - name: CodeFileSanity
+          run: |
+            # Suppress warnings from templates project
+            # https://unix.stackexchange.com/questions/104641/how-can-i-suppress-output-from-grep-so-that-it-only-returns-the-exit-status
+            dotnet codefilesanity | while read -r line; do
+               if [ $(echo $line| grep -q "./osu.Framework.Templates/"; echo $?) == 1 ]; then
+                 echo "::warning::$line"
+               fi
+            done
+
+        - name: .NET Format (Dry Run)
+          run: dotnet format --dry-run --check
+
+        - name: InspectCode
+          run: |
+            dotnet jb inspectcode $(pwd)/osu-framework.Desktop.slnf --output=inspectcodereport.xml --cachesDir=$(pwd)/inspectcode --verbosity=WARN
+            dotnet nvika parsereport $(pwd)/inspectcodereport.xml  --treatwarningsaserrors
```

Have you tested this also fails the CI process? Nvika's quite old (came before gh-actions) so I'd be surprised if it does (also: https://github.com/laedit/vika/issues/224).

```suggestion
            dotnet nvika parsereport $(pwd)/inspectcodereport.xml --treatwarningsaserrors
```

---

Haven't introduced a bad change to test things out, but Actions is designed to error if the said workflow sends a non-zero exit code (which is fairly standard for most analyzers).

---

Yeah I know that about analyzers exiting on non-zero, wasn't sure whether nvika did that. Seems like it does:
```
[smgi@smgi osu-framework]$ dotnet jb inspectcode osu-framework.Desktop.slnf --output="build/inspectcodereport.xml" --verbosity=WARN && dotnet nvika parsereport build/inspectcodereport.xml --treatwarningsaserrors

...

[Fatal] Issues with severity error was found: the build will fail

[smgi@smgi osu-framework]$ $?
-bash: 5: command not found
```

We also can't just have it error and have the user fish for it in plain text. The inspectcode command already dumps thousands of lines of output (one for each file).

How does it look when it fails, can you provide a screenshot?

---

> We also can't just have it error and have the user fish for it in plain text. The inspectcode command already dumps thousands of lines of output (one for each file).

Hmm, since InspectCode is also a GitHub Action, I think it'll be much safer to use that instead? [I think this is pretty close but I doubt this is what we want.](https://github.com/glassechidna/resharper-action)

---

It's slightly out of date but you may be able to use the xml option (rather than solution):
```
with:
    report: $(pwd)/inspectcodereport.xml
```

---

This should be resolved now by using the ReSharper Action.